### PR TITLE
Technical advisory 79281

### PR DESCRIPTION
**Pull request by MichaelTrestman**
https://github.com/cockroachdb/cockroach/pull/79293
https://github.com/cockroachdb/cockroach/issues/79281

### CODE REVIEW

File: advisories/a79281.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+---
+title: Technical Advisory 79281
+summary: Importing duplicate keys can cause violations of UNIQUE constrants
```
Comment (kathancox) [inline advisories/a79281.md:3]: Typo here: "constraints"

File: advisories/a79281.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+## Mitigation
+
+Users of CockroachDB v21.2.0 to v21.2.7 who have used [`IMPORT INTO`](../v22.1/import-into.html) should audit their [`REGIONAL BY ROW`](../v22.1/regional-tables.html) tables with `UNIQUE` constraints to ensure these tables do not contain `UNIQUE` constraint violations.
+
+Users should also upgrade to v21.2.8 or a later version.
+
+### v21.2.6 or later
+
+Users of CockroachDB v21.2.6 and later have access to the following [built-in functions](../v22.1/functions-and-operators.html), which can be used to easily validate unique constraints:
```
Comment (kathancox) [inline advisories/a79281.md:31]: Maybe cap and inline code on "unique" here to match elsewhere?

File: advisories/a79281.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+- `SELECT crdb_internal.revalidate_unique_constraints_in_all_tables();`
+- `SELECT crdb_internal.revalidate_unique_constraints_in_table('table_name');`
+- `SELECT crdb_internal.revalidate_unique_constraint('table_name', 'constraint_name');`
+
+If any constraint fails validation, the functions will return an error with a hint about which data caused the constraint violation. These violations can then be resolved manually by updating or deleting the rows in violation.
+
+### Prior to v21.2.6
+
+Users of CockroachDB releases prior to v21.2.6 may audit the table for duplicate keys by running a query of the following form each [`UNIQUE`](../v22.1/unique.html) constraint:
```
Comment (kathancox) [inline advisories/a79281.md:41]: I think this is meant to say "... a query of the following for each..." Maybe a typo on "form"

File: advisories/a79281.md (hunk 2 above)
Comment (kathancox) [reply advisories/a79281.md:31]: And there is a later one too, so perhaps a quick double check on these in the file?

File: advisories/a79281.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+SELECT * FROM t GROUP BY unique_col1 [, unique_col2 ...] HAVING count(*) > 1;
+~~~
+
+If a query returns any results, this indicates that the table contains unique constraint violations. To fix the violations, update all but one of the duplicate keys with a different value.
+
+## Impact
+
+[`REGIONAL BY ROW`](../v22.1/regional-tables.html) tables spanning multiple regions could contain non-unique data in unique columns if [`IMPORT INTO`](../v22.1/import-into.html) was used to import data with duplicate keys. Versions affected include v21.2.0 to v21.2.7, v22.1.0-alpha.1 to v22.1.0-alpha.5, and v22.1.0-beta.1.
+
+Please reach out to the [support team](https://support.cockroachlabs.com/) if more information or assistance is needed.
```
Comment (kathancox) [inline advisories/a79281.md:53]: Typo on "if" here, I think it should be "for":  
"... reach out to the support team for more information"

**Review by kathancox — COMMENTED** [review]
This looks good to me! A few v minor nits.

**Review by MichaelTrestman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/13582)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kathancox)

___
*[advisories/a79281.md, line 3 at r1](https://reviewable.io/reviews/cockroachdb/docs/13582#-N-Q4ivEs3sVmaeRhsBI-r1-3:-N-QF0ne8H1DdKzWowd4:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cc40744999dedda57fdd5bcd66e66984683f44c7/advisories/a79281.md#L3)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

Typo here: "constraints"
</blockquote></details>

Done.
___
*[advisories/a79281.md, line 31 at r1](https://reviewable.io/reviews/cockroachdb/docs/13582#-N-Q4ivEs3sVmaeRhsBI-r1-31:-N-QF7PB-QYbWaixSxBr:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cc40744999dedda57fdd5bcd66e66984683f44c7/advisories/a79281.md#L31)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

And there is a later one too, so perhaps a quick double check on these in the file?
</blockquote></details>

Done.
___
*[advisories/a79281.md, line 41 at r1](https://reviewable.io/reviews/cockroachdb/docs/13582#-N-Q4ivEs3sVmaeRhsBI-r1-41:-N-QFX0Q1IN0UciED461:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cc40744999dedda57fdd5bcd66e66984683f44c7/advisories/a79281.md#L41)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

I think this is meant to say "... a query of the following for each..." Maybe a typo on "form" 
</blockquote></details>

Done.
___
*[advisories/a79281.md, line 53 at r1](https://reviewable.io/reviews/cockroachdb/docs/13582#-N-Q4ivEs3sVmaeRhsBI-r1-53:-N-QFpod721-zT86N3v-:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/cc40744999dedda57fdd5bcd66e66984683f44c7/advisories/a79281.md#L53)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

Typo on "if" here, I think it should be "for":  
"... reach out to the support team for more information"
</blockquote></details>

Done.

**Comment by kathancox** [comment]
This looks good to me from docs, wasn't sure if there should be any engineer or TA process people reviewing before publishing?

**Review by kathancox — APPROVED** [review]

**Pull request merged**